File: taesunpark/2021/1108~1114/징검다리.py
def solution(distance, rocks, n):
    answer = 0
    rocks.sort()
    start = 0
    end = distance
    while start <= end:
        mid = (start + end) // 2  # mid는 가장 작은 거리임
        del_stone = 0  # 삭제된 돌 수
        loc_stone = 0  # 돌 위치
        for i in rocks:
            if i - loc_stone < mid:
                del_stone += 1
            else:
                loc_stone = i
            if del_stone > n:
                break;

        if del_stone > n:
            end = mid - 1
        else:
            answer = mid
            start = mid + 1

    return answer

File: hyeon-uk/2021/1108~1114/징검다리.cpp
#include <string>
#include <vector>
#include<algorithm>

using namespace std;

int solution(int distance, vector<int> rocks, int n) {
    int answer = 0;
    rocks.push_back(distance);
    
    sort(rocks.begin(),rocks.end());
    
    int left=0;
    int right=1000000000;
    
    while(left<=right){
        int mid=(left+right)/2;
        int sum=0;
        int st=0;
        for(int i=0;i<rocks.size();i++){
            if(rocks[i]-st<mid){
                sum++;
            }
            else{
                st=rocks[i];
            }
        }
        
        if(sum>n){
            right=mid-1;
        }
        else{
            answer=mid;
            left=mid+1;
        }
    }
    return answer;
}
